[identity] Added AzurePowerShellCredential

witemple-msft opened the pull request:
This is something we implemented in the official SDK packages for authentication through Azure PowerShell.

Couple of notes:
- It's added at the end of `DefaultAzureCredential` sources because it's slower than the others (by a lot)
- It has a static `new` function that accepts an optional PathBuf for the path of the powershell executable to use. 
- If the path option wasn't given during construction, it tries `pwsh.exe` and then `powershell.exe` on Windows. On non-windows it just tries to execute `pwsh`. I've used a lazy_static to make it only try this dance once.
- I'm using `tokio::process` to call PowerShell instead of std::process. I figure this is slightly preferable, but does require adding tokio to the deps instead of just dev-deps.
- I chose to use MS graph in the example instead of storage accounts just because I think it shows a very clear confirmation of _who_ you're authenticated as. I'm not 100% sure of what it will output if you connect a service principal instead of a user principal.

It's been a while since I've written a much Rust regularly, so any feedback on the style/implementation is very much appreciated, but I might as well try to pitch in some of the enhancements I've worked on for the TypeScript SDK.

rylev commented on sdk/identity/src/token_credentials/powershell_credential.rs:
```diff
@@ -0,0 +1,179 @@
+#[non_exhaustive]
+#[derive(Debug, thiserror::Error)]
+pub enum AzurePowerShellCredentialError {
+    #[error("Failed to import the Az.Account module (is it installed?): {0}")]
```
I've done this in #306 - can you make the first letter of all the error messages lowercase? This is how the error handling working group recommends messages are formatted.

rylev commented on sdk/identity/src/token_credentials/powershell_credential.rs:
```diff
@@ -0,0 +1,179 @@
+
+fn error_to_powershellcredential_error<T>(
+    error: Error,
+) -> Result<T, AzurePowerShellCredentialError> {
```
Hmm why does this return a `Result`? Can this just be a `From<Error> for AzurePowerShellCredentialError`?

rylev commented on sdk/identity/src/token_credentials/powershell_credential.rs:
```diff
@@ -0,0 +1,179 @@
+}
+
+impl AzurePowerShellCredential {
+    pub fn new(powershell_path: Option<PathBuf>) -> AzurePowerShellCredential {
```
nit: I think we're preferring `-> Self` and `Self {` instead of renaming the type.

rylev commented on sdk/identity/src/token_credentials/powershell_credential.rs:
```diff
@@ -0,0 +1,179 @@
+            Ok(response) if !response.status.success() => {
+                let output = String::from_utf8_lossy(&response.stderr);
+                return Err(AzurePowerShellCredentialError::ImportFailed(
+                    output.to_string(),
```
the call to `to_string` is not necessary

witemple-msft replied:
Hm, `from_utf8_lossy` returns a `Cow<str>`. I think the conversion to String is required here and FWIW AzureCliCredential does the same in this case. The other option is that maybe I don't really need to use lossy conversion in this case, but figured it was simpler to do that than handle an Err from potentially invalid UTF-8 in the error output of the command.

rylev replied:
Ah yes - sorry. `to_owned` might be clearer (and potentially every so slightly more performant), but you don't have to change it if you don't want.

rylev commented on sdk/identity/src/token_credentials/powershell_credential.rs:
```diff
@@ -0,0 +1,179 @@
+            _ => { /* Everything else is OK to proceed. */ }
+        };
+
+        let resource_fragment = match resource {
+            Some(scope) => format!("-ResourceUrl \"{}\" ", get_scope_resource(scope)),
+            None => String::from(""),
+        };
```
Nit:
```suggestion
        let resource_fragment = resource.map(|scope| format!("-ResourceUrl \"{}\" ", get_scope_resource(scope))).unwrap_or_default();
```

rylev commented on sdk/identity/src/token_credentials/powershell_credential.rs:
```diff
@@ -0,0 +1,179 @@
+            Ok(get_token_output) => {
+                let output = String::from_utf8_lossy(&get_token_output.stderr);
+                Err(AzurePowerShellCredentialError::CommandFailed(
+                    output.to_string(),
```
Remove `to_string`

rylev reviewed (changes requested):
Looking great! 🎉 

Let's switch to `std::process::Process`, threads, and an async channel for communication for now. There should be async channel implementations we can use that don't depend directly on a runtime.

witemple-msft commented:
@rylev I took a look at using threads/async_channel instead of the tokio runtime this morning. It works well, but one of the consequences is that I did have to do a bit of shuffling around to make sure that everything we need to start the `std::Process::Command` can be moved into the thread's closure (`Into<OsString>` instead of `AsRef<OsStr>`, etc.)

Addressed all the other points and/or left comments above.

rylev commented on sdk/identity/src/token_credentials/powershell_credential.rs:
```diff
@@ -0,0 +1,190 @@
+        I: IntoIterator<Item = S>,
+        S: Into<OsString>,
+    {
+        let pwsh = self.powershell_path.as_ref().unwrap_or(&*POWERSHELL_PATH);
```
Won't this cause `get_default_powershell_path` to run on the current thread? I think we want this also to run on a separate thread no?

rylev replied:
Also, you'll want to use `unwrap_or_else` so that `POWERSHELL_PATH` is only evaluated if `self.powershell_path` is `None`.

rylev commented on sdk/identity/src/token_credentials/powershell_credential.rs:
```diff
@@ -0,0 +1,190 @@
+            ])
+            .await?;
+
+        if !get_token_response.status.success() {
+            let output = String::from_utf8_lossy(&get_token_response.stderr);
+            Err(AzurePowerShellCredentialError::CommandFailed(
+                output.to_string(),
+            ))
+        } else {
```
Nit: I prefer the success case to be first before the non-success case.

rylev reviewed (changes requested):
Looks good. I'd just like to clarify on what thread the call to find out the powershell path will be run.

witemple-msft commented:
@rylev Sorry it took so long, but I've addressed the question with the latest changes. Got caught up with work and moving to a new place :sweat_smile:. I switched the initialization logic from lazy_static to async_once_cell and made the windows path initialization code that tests for `pwsh` run on a separate thread.

rylev reviewed (approved):
Thanks! 🎉 This looks good to me, but I'd like to get some other eyes on this as well.

cataggar commented:
Please reopen when this is active again.
